client/messaging/message-list.tsx

```diff
@@ -41,34 +35,44 @@ const Messages = styled.div`
 `
 
 interface PureMessageListProps {
-  messages: List<ChatMessage>
+  messages: List<Message>
+  renderMessages?: (msg: Message) => ReactNode
 }
 
-function messageToElem(msg: ChatMessage) {
+function messageToElem(msg: CommonMessage) {
   switch (msg.type) {
-    case ChatMessageType.JoinChannel:
-      return <JoinChannelMessage key={msg.id} record={msg} />
-    case ChatMessageType.LeaveChannel:
-      return <LeaveChannelMessage key={msg.id} record={msg} />
-    case ChatMessageType.TextMessage:
+    case CommonMessageType.TextMessage:
       return <TextMessageDisplay key={msg.id} user={msg.from} time={msg.time} text={msg.text} />
-    case ChatMessageType.NewChannelOwner:
-      return <NewChannelOwnerMessage key={msg.id} record={msg} />
-    case ChatMessageType.SelfJoinChannel:
-      return <SelfJoinChannelMessage key={msg.id} record={msg} />
     default:
       return null
   }
 }
 
 // This contains just the messages, to avoid needing to re-render them all if e.g. loading state
 // changes on the actual message list
-const PureMessageList = React.memo<PureMessageListProps>(({ messages }) => {
-  return <Messages>{messages.map(m => messageToElem(m))}</Messages>
+const PureMessageList = React.memo<PureMessageListProps>(({ messages, renderMessages }) => {
+  return (
+    <Messages>
+      {messages.map(m => {
+        // NOTE(2Pac): We only handle common messages here, e.g. text message. All other types of
+        // messages are handled by calling the `renderMessages` function below which should be
+        // supplied by each service if they have any special messages to handle.
+        if (Object.values(CommonMessageType).includes(m.type as CommonMessageType)) {
```

Comment:
collecting the values and doing an includes on each message seems kind of expensive. At the very least I would collect the values into a readonly array outside this function (or a Set for a probably-faster inclusion check), but honestly I think a simpler approach is just:

```ts
return messageToElem(m) ?? (renderMessage ?? DEFAULT_RENDER_MESSAGE)(m)
```